### PR TITLE
Optimize RoPE and Attention

### PR DESCRIPTION
**Pull request by lapp0**
# Overview

Numerically equivalent optimizations, no quantization.

- **RoPE**: Don't store frequency buffers, precompute audio/video sin/cos buffers
- **Attention Masking**: No for loops to construct attention mask, use `flex_attention` / `BlockMask`. Enables `torch.compile`

|                          | Forward (ms) | Backward (ms) |
|--------------------------------|--------------|---------------|
| baseline                       | 66.358       | 365.601       |
| optimize RoPE                  | 49.324       | 333.734       |
| flex_attention BlockMask  + compile      |    44.811   |   125.035     |

Note: 1xH100 SXM, using `av_v2_4x4.yml` DiT model with 12 layers

# Integrity checks

Output for `test_attn_mask`:

<details>
image
</details>

RoPE includes a checksum test which ensures numerical equivalence

## Baseline

[Profile](https://wandb.ai/lapp0-andrew-lapp/owl/artifacts/profile/trace/v12/files/e952766aad7c_6164.1752522863635627798.pt.trace.json)

<details>
image
</details>

## RoPE

[Profile](https://wandb.ai/lapp0-andrew-lapp/owl/artifacts/profile/trace/v13/files/e952766aad7c_6805.1752523588957360194.pt.trace.json)

<details>
image
</details>

## flex_attention + compile

Now that attn mask construction has no for loop, we can compile the attention module

[Profile](https://wandb.ai/lapp0-andrew-lapp/owl/artifacts/profile/trace/v17/files/e952766aad7c_174.1752531331662015072.pt.trace.json)

# Future work
- **RoPE**: merge the four audio cos, audio sin, video cos, video sin buffers into only two `cos` / `sin` buffers - (estimated 2x improvement of RoPE performance)
- **Attention Mask** (inference only) - compute once, dynamically update BlockMask when the kv cache updates. Current behavior is to construct a new BlockMask based on the kv cache size for every forward pass.
- **Compile**: we can trivially compile the entire model, not just `Attn` as a next step 
- **cuDNN Attention**: Better than flash attention in all other models I've worked with. (trivial: `enable_cudnn_sdp(True); enable_flash_sdp(False)`)

### CODE REVIEW

File: owl_wms/nn/attn.py [hunk 1]
```diff
@@ -129,22 +120,43 @@ class DiT(nn.Module):
...
         for i in range(config.n_layers):
             blocks.append(DiTBlock(config))
             blocks[-1].attn.layer_ind = i
         self.blocks = nn.ModuleList(blocks)
 
+    def get_block_mask(self, x, kv_cache):
+        if not self.causal:
+            return None
+        B, L, _ = x.shape
+        offset = kv_cache.length_at(0) if kv_cache is not None else 0
```
Comment (lapp0) [inline owl_wms/nn/attn.py:136]: Assumes all layers have same kv cache size

File: owl_wms/nn/attn.py [hunk 2]
```diff
@@ -50,51 +60,32 @@ def __init__(self, config : 'TransformerConfig'):
...
-            new_v = torch.cat([old_v, v], dim = 2).contiguous()
-
+            k = torch.cat([old_k, k], dim=2)
+            v = torch.cat([old_v, v], dim=2)
             if kv_cache.should_update:
-                kv_cache.update(new_k.clone(), new_v.clone(), self.layer_ind)
+                kv_cache.update(k.clone(), v.clone(), self.layer_ind)
 
-            q,new_k = self.rope(q, new_k)
+        q, k = self.rope(q, k)
```
[the remark this answers is not in the document]
Comment (LouisCastricato) [reply owl_wms/nn/attn.py:82]: getting this for you right now.

File: owl_wms/nn/attn.py [hunk 3]
```diff
@@ -50,51 +60,32 @@ def __init__(self, config : 'TransformerConfig'):
...
-        )
-    
-    def forward(self, x, kv_cache = None):
+    @torch.compile
+    def forward(self, x, block_mask, kv_cache=None):
+        B, L, _ = x.shape
+
         qkv = self.qkv(x)
         qkv = qkv.view(qkv.shape[0], qkv.shape[1], 3, self.n_heads, -1)
         q, k, v = qkv.permute(2, 0, 3, 1, 4)
```
Comment (LouisCastricato) [inline owl_wms/nn/attn.py:69]: Can you change this to an einops?

Comment (LouisCastricato) [reply owl_wms/nn/attn.py:69]: no big deal if not.

File: owl_wms/nn/attn.py (hunk 1 above)
Comment (LouisCastricato) [reply owl_wms/nn/attn.py:136]: They do

File: owl_wms/nn/rope.py [hunk 4]
```diff
@@ -11,103 +11,66 @@
 from torch import nn
 import time
 
+
 class FlatVideoRoPE(nn.Module):
```
Comment (LouisCastricato) [inline owl_wms/nn/rope.py:15]: Need to make sure this works the same way once shahbuland provides you checkpoints.

File: owl_wms/nn/rope.py [hunk 5]
```diff
@@ -11,103 +11,66 @@
...
+        aud_ang = RotaryEmbedding(d_head // 2)\
+            .get_axial_freqs(config.n_frames)
+
+        self.register_buffer("vcos", vid_ang.cos(), persistent=False)
+        self.register_buffer("vsin", vid_ang.sin(), persistent=False)
+        self.register_buffer("acos", aud_ang.cos(), persistent=False)
+        self.register_buffer("asin", aud_ang.sin(), persistent=False)
+
+    @staticmethod
+    def _rot(x: torch.Tensor, cos: torch.Tensor, sin: torch.Tensor) -> torch.Tensor:
```
Comment (LouisCastricato) [inline owl_wms/nn/rope.py:38]: Add back some of the original comments.

File: owl_wms/nn/attn.py (hunk 3 above)
Comment (lapp0) [reply owl_wms/nn/attn.py:69]: Making a note to do something like the below once I'm done testing
```
q, k, v = rearrange(qkv, "b t (3 h d) -> 3 b h t d", h=self.n_heads)
```
